lib/strings.nix

```
@@ -238,7 +243,12 @@ rec {
     # Prefix to check for
     pref:
     # Input string
-    str: substring 0 (stringLength pref) str == pref;
+    str:
+    # pref must be a string
+    # str must be a string-like
+    if isPath pref then throw ''
+      lib.strings.hasPrefix: The first argument (${toString pref}) is a path, which is not supported''
```

```suggestion
    # Before 23.05, strings would be added to the store before comparing. This was too unexpected.
    if isPath pref then throw ''
      lib.strings.hasPrefix: The first argument (${toString pref}) is a path value, but only strings are supported.''
```

Leave a clue in case users encounter this in their code base, so they know something has changed here. It appears in the code block of the stack trace iirc. The exception message does not include the whole history, because I don't expect that to be productive. It could stick around for years being unnecessarily noisy.

---

Sounds good, but "only strings are supported" isn't entirely correct, all `stringLike` values (including `outPath`, `__toString`) are supported, except paths. Maybe that's good enough though

---

Right. Probably best to keep it simple and avoid unnecessary dissonance?
This can only be cured with poetry. Cue ChatGPT:

Here is a revised limerick about type coercions in programming:

Type coercion can be quite slick,
To convert data types that don't mix.
But if you're not careful,
Your code can be awful,
And bugs will give your code a big kick.

---

Haha nicely done :laughing:

I agree, let's keep it simple, I applied this suggestion to all the error messages (also the normalizePath one).